packages/auth/README.md

```diff
@@ -307,33 +308,288 @@ console.log(context.currentUser)
 
 You can map the "raw decoded JWT" into a real user object by passing a `getCurrentUser` function to `createCreateGraphQLHandler`
 
-Our recommendation is to create a `src/lib/auth.js|ts` file that exports a `getCurrentUser` function (You may already have a stub function.)
+Our recommendation is to create a `src/lib/auth.js|ts` file that exports a `getCurrentUser` as well as a `hasRole` function (You may already have stub functions.)
 
 ```js
-import { getCurrentUser } from 'src/lib/auth'
+import { getCurrentUser, hasRole } from 'src/lib/auth'
 // Example:
 //  export const getCurrentUser = async (decoded) => {
 //    return await db.user.findOne({ where: { decoded.email } })
 //  }
-``
+//
+// Example for hasRole and RBAC
+//
+//  const appMetadata = (decoded) => {
+//    return decoded[`https://example.com/app_metadata`] || {}
+//  }
+//
+//  const roles = (decoded) => {
+//    return appMetadata(decoded).authorization?.roles || []
+//  }
+//
+//  export const getCurrentUser = async (decoded) => {
+//    const user = await db.user.findOne({ where: { decoded.email } })
+//    return { ...user, roles: roles(decoded) }
+//  }
+//
+//  export const hasRole = (role) => {
+//    if (!context.currentUser.roles?.includes(role)) {
+//      throw new AuthenticationError("You don't have permission to do that.")
+//    }
+//  }
+
 
 export const handler = createGraphQLHandler({
   schema: makeMergedSchema({
     schemas,
     services: makeServices({ services }),
   }),
   getCurrentUser,
+  hasRole, // if role access
 })
 ```
 
 The value returned by `getCurrentUser` is available in `context.currentUser`
 
+The `hasRole` function is generated and defined in `api/src/lib/auth.js`
+
+```js
+export const hasRole = (role) => {
+  if (!context.currentUser.roles?.includes(role)) {
+    throw new AuthenticationError("You don't have permission to do that.")
```

Comment:
Suggestion: create a new type of error, maybe AuthorizationError? (not sure how to just yet)

This is helpful, for example, because we're using "AuthenticationError" to redirect the user to the login page

Reply:
@dac09 I like the `AuthorizationError` idea. I *think* `AuthenticationError` comes from [Apollo Server](https://www.apollographql.com/docs/apollo-server/data/errors/) and there looks to be a `ForbiddenError` that makes more sense to use.

And I think `ForbiddenError` can be imported from `redwoodjs/api` like `AuthenticationError` is.

So ...

```js
import { AuthenticationError, ForbiddenError } from '@redwoodjs/api'
...

throw new ForbiddenError("You don't have access to do that.")
```

Or some much better message.

Reply:
Yup - that works:

```js
import { AuthenticationError, ForbiddenError } from '@redwoodjs/api'

// ....

export const hasRole = (role) => {
  if (!context.currentUser.roles?.includes(role)) {
    throw new ForbiddenError("You don't have access to do that.")
  }
}
```

and when protecting a service the error output of a cell is:

```
Error: GraphQL error: You don't have access to do that.
```